File: sdk/java/src/main/java/com/xxxxteam/tempmail/providers/TempEmailCo.java
package com.xxxxteam.tempmail.providers;

import com.google.gson.JsonElement;
import com.xxxxteam.tempmail.Email;
import com.xxxxteam.tempmail.EmailInfo;
import com.xxxxteam.tempmail.HttpClient;
import com.xxxxteam.tempmail.HttpResult;
import com.xxxxteam.tempmail.Json;
import com.xxxxteam.tempmail.Normalizer;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * tempemail.co 渠道 — https://tempemail.co
 * GET /mail/random 创建，GET /get-mails 获取邮件列表（HTML regex 提取 ID），
 * GET /mail/info?id={id} 获取邮件详情。token 为 JSON {"address":"...","cookies":"..."}。
 */
public final class TempEmailCo {

    private static final String BASE_URL = "https://tempemail.co";
    private static final Pattern MAIL_ID_RE = Pattern.compile("data-id=\"(\\d+)\"");
    private static final String UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            + "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0";

    private TempEmailCo() {
    }

    private static Map<String, String> jsonHeaders(String cookie) {
        Map<String, String> h = new LinkedHashMap<>();
        h.put("User-Agent", UA);
        h.put("Accept", "application/json, text/javascript, */*; q=0.01");
        h.put("Referer", BASE_URL + "/");
        if (cookie != null && !cookie.isEmpty()) {
            h.put("Cookie", cookie);
        }
        return h;
    }

    /**
     * 创建临时邮箱。
     *
     * @return 邮箱信息，token 为 JSON {"address":"...","cookies":"..."}
     */
    public static EmailInfo generate() {
        HttpResult resp = HttpClient.get(BASE_URL + "/mail/random", jsonHeaders(""));
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        JsonElement resultEl = data.getAsJsonObject().get("result");
        if (resultEl == null || !resultEl.getAsBoolean()) {
            throw new RuntimeException("tempemail-co: 创建邮箱失败");
        }
        String address = ProviderUtil.firstNonEmpty(
                Json.str(data, "address"), Json.str(data, "id")).trim();
        if (address.isEmpty() || !address.contains("@")) {
            throw new RuntimeException("tempemail-co: 邮箱地址无效");
        }
        String cookie = ProviderUtil.extractAllCookies(resp);
        String token = "{\"address\":\"" + address + "\",\"cookies\":\"" + cookie.replace("\"", "\\\"") + "\"}";
        return new EmailInfo("tempemail-co", address, token, null, null);
    }

    /**
     * 获取邮件列表。
     *
     * @param token JSON 格式认证信息
     * @param email 邮箱地址
     * @return 邮件列表
     */
    public static List<Email> getEmails(String token, String email) {
        String address = (email != null ? email : "").trim();
        if (address.isEmpty()) {
            throw new RuntimeException("tempemail-co: 邮箱地址为空");
        }
        JsonElement tokenData = Json.parse(token != null ? token : "{}");
        String storedAddr = Json.str(tokenData, "address").trim();
        String cookie = Json.str(tokenData, "cookies").trim();
        String queryAddr = storedAddr.isEmpty() ? address : storedAddr;
        // GET /get-mails
        String url = BASE_URL + "/get-mails?mail_id=" + ProviderUtil.urlEncode(queryAddr)
                + "&unseen=0&is_new=1";
        HttpResult resp = HttpClient.get(url, jsonHeaders(cookie));
        resp.ensureSuccess();
        JsonElement data = Json.parse(resp.getBody());
        JsonElement countEl = data.getAsJsonObject().get("count");
        int count = 0;
        if (countEl != null && countEl.isJsonPrimitive()) {
            try { count = countEl.getAsInt(); } catch (Exception ignored) {}
        }
        if (count <= 0) {
            return new ArrayList<>();
        }
        // 从 mails HTML 中提取邮件 ID
        String mailsHtml = Json.str(data, "mails");
        if (mailsHtml.isEmpty()) {
            return new ArrayList<>();
        }
        Set<String> uniqueIds = new LinkedHashSet<>();
        Matcher m = MAIL_ID_RE.matcher(mailsHtml);
        while (m.find()) {
            uniqueIds.add(m.group(1));
        }
        if (uniqueIds.isEmpty()) {
            return new ArrayList<>();
        }
        List<Email> result = new ArrayList<>();
        for (String mailId : uniqueIds) {
            Map<String, Object> raw = fetchMailInfo(cookie, mailId, address);
            if (raw != null) {
                result.add(Normalizer.normalizeEmail(raw, address));
            }
        }
        return result;
    }

    private static Map<String, Object> fetchMailInfo(String cookie, String mailId, String recipient) {
        try {
            String url = BASE_URL + "/mail/info?id=" + ProviderUtil.urlEncode(mailId);
            HttpResult resp = HttpClient.get(url, jsonHeaders(cookie));
            if (!resp.isOk()) return null;
            JsonElement data = Json.parse(resp.getBody());
            JsonElement resultEl = data.getAsJsonObject().get("result");
            if (resultEl == null || !resultEl.getAsBoolean()) return null;
            JsonElement mailEl = data.getAsJsonObject().get("mail");
            if (mailEl == null || !mailEl.isJsonObject()) return null;
            String fromName = Json.str(mailEl, "fromName").trim();
            String fromAddr = Json.str(mailEl, "fromAddress").trim();
            String fromDisplay = fromName.isEmpty() ? fromAddr : fromName + " <" + fromAddr + ">";
            Map<String, Object> raw = new LinkedHashMap<>();
            raw.put("id", mailId);
            raw.put("from", fromDisplay);
            raw.put("from_address", fromAddr);
            raw.put("to", recipient);
            raw.put("subject", Json.str(mailEl, "subject"));
            raw.put("html", Json.str(mailEl, "textHtml"));
            raw.put("date", Json.str(mailEl, "displayDate"));
            return raw;
        } catch (Exception e) {
            return null;
        }
    }
}

File: sdk/kotlin/src/main/kotlin/com/xxxxteam/tempmail/providers/TempEmailCo.kt
package com.xxxxteam.tempmail.providers

import com.xxxxteam.tempmail.Email
import com.xxxxteam.tempmail.EmailInfo
import com.xxxxteam.tempmail.Normalize
import com.xxxxteam.tempmail.Provider
import kotlinx.serialization.json.*

/**
 * tempemail.co 渠道实现 — https://tempemail.co
 *
 * GET /mail/random 创建，GET /get-mails 获取邮件列表（HTML regex 提取 ID），
 * GET /mail/info?id={id} 获取邮件详情。
 * token 为 JSON {"address":"...","cookies":"..."}。
 */
object TempEmailCo : Provider {

    private const val BASE_URL = "https://tempemail.co"
    private val MAIL_ID_RE = Regex("data-id=\"(\\d+)\"")
    private const val UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"

    private fun jsonHeaders(cookie: String): Map<String, String> {
        val h = linkedMapOf(
            "User-Agent" to UA,
            "Accept" to "application/json, text/javascript, */*; q=0.01",
            "Referer" to "$BASE_URL/",
        )
        if (cookie.isNotEmpty()) h["Cookie"] = cookie
        return h
    }

    /** 创建临时邮箱。 */
    override suspend fun generate(): EmailInfo {
        val resp = ProviderUtil.httpGet("$BASE_URL/mail/random", jsonHeaders(""))
        resp.ensureSuccess()
        val data = ProviderUtil.parseObject(resp.body)
        if (data == null || !ProviderUtil.bool(data, "result")) {
            throw RuntimeException("tempemail-co: 创建邮箱失败")
        }
        val address = ProviderUtil.firstNonEmpty(
            ProviderUtil.str(data, "address"), ProviderUtil.str(data, "id")).trim()
        if (address.isEmpty() || !address.contains("@")) {
            throw RuntimeException("tempemail-co: 邮箱地址无效")
        }
        val cookie = ProviderUtil.extractAllCookies(resp)
        val token = buildJsonObject {
            put("address", address)
            put("cookies", cookie)
        }.toString()
        return EmailInfo(email = address, channel = "tempemail-co", token = token)
    }

    /** 获取邮件列表。 */
    override suspend fun getEmails(info: EmailInfo): List<Email> {
        val address = info.email.trim()
        if (address.isEmpty()) throw RuntimeException("tempemail-co: 邮箱地址为空")
        val tokenData = ProviderUtil.parseObject(info.token.ifBlank { "{}" })
        val storedAddr = ProviderUtil.str(tokenData, "address").trim()
        val cookie = ProviderUtil.str(tokenData, "cookies").trim()
        val queryAddr = storedAddr.ifEmpty { address }
        val url = "$BASE_URL/get-mails?mail_id=${ProviderUtil.urlEncode(queryAddr)}&unseen=0&is_new=1"
        val resp = ProviderUtil.httpGet(url, jsonHeaders(cookie))
        resp.ensureSuccess()
        val data = ProviderUtil.parseObject(resp.body) ?: return emptyList()
        val count = (data["count"] as? JsonPrimitive)?.content?.toIntOrNull() ?: 0
        if (count <= 0) return emptyList()
        val mailsHtml = ProviderUtil.str(data, "mails")
        if (mailsHtml.isEmpty()) return emptyList()
        val uniqueIds = LinkedHashSet<String>()
        MAIL_ID_RE.findAll(mailsHtml).forEach { uniqueIds.add(it.groupValues[1]) }
        if (uniqueIds.isEmpty()) return emptyList()
        val result = ArrayList<Email>(uniqueIds.size)
        for (mailId in uniqueIds) {
            val raw = fetchMailInfo(cookie, mailId, address)
            if (raw != null) result.add(Normalize.fromMap(raw, address))
        }
        return result
    }

    /** 拉取单封邮件详情，失败返回 null。 */
    private suspend fun fetchMailInfo(cookie: String, mailId: String, recipient: String): Map<String, String>? {
        return try {
            val resp = ProviderUtil.httpGet(
                "$BASE_URL/mail/info?id=${ProviderUtil.urlEncode(mailId)}", jsonHeaders(cookie))
            if (!resp.isOk) return null
            val data = ProviderUtil.parseObject(resp.body) ?: return null
            if (!ProviderUtil.bool(data, "result")) return null
            val mail = data["mail"] as? JsonObject ?: return null
            val fromName = ProviderUtil.str(mail, "fromName").trim()
            val fromAddr = ProviderUtil.str(mail, "fromAddress").trim()
            val fromDisplay = if (fromName.isEmpty()) fromAddr else "$fromName <$fromAddr>"
            mapOf(
                "id" to mailId,
                "from" to fromDisplay,
                "from_address" to fromAddr,
                "to" to recipient,
                "subject" to ProviderUtil.str(mail, "subject"),
                "html" to ProviderUtil.str(mail, "textHtml"),
                "date" to ProviderUtil.str(mail, "displayDate"),
            )
        } catch (_: Exception) {
            null
        }
    }
}
